File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include<iostream>
#include<string>
#include<vector>
#include<cstring>
#include<algorithm>
#include<queue>
#include<utility>
#include<cmath>
using namespace std;
typedef pair<int, int>p;
int ans = 2e9, arr[50][50];
vector<p>chicken,home;
vector<int>dist;
int selected[13];
int n, m, blank;
vector<p>q;
void input() {
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> arr[i][j];
			if (arr[i][j] == 2) chicken.push_back({ i,j });
			if (arr[i][j] == 1) home.push_back({ i,j });
		}
	}
	dist.resize(home.size());
}
void go() {
	int sum = 0;
	for(int i=0;i<home.size();i++) {
		int dist = 2e9;
		for (int j = 0; j < q.size(); j++) {
			p now = q[j];
			dist = min(dist, abs(now.first - home[i].first) + abs(now.second - home[i].second));
		}
		sum += dist;
	}
	ans = min(ans, sum);

}
void combination(int idx, int cnt) {
	if (cnt == m) {
		q.clear();
		for (int i = 0; i < chicken.size(); i++) {
			if (selected[i]) {
				int x = chicken[i].first, y = chicken[i].second;
				q.push_back({ x,y });
			}
		}
		go();
		return;
	}
	for (int i = idx; i < chicken.size(); i++) {
		if (!selected[i]) {
			selected[i] = 1;
			combination(i + 1, cnt + 1);
			selected[i] = 0;
		}
	}

}
int main() {
	ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
	input();
	combination(0, 0); //조합
	cout << ans << "\n";
	return 0;
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
	static int ans = (int) 2e9;
	static boolean selected[];
	static int n, m;
	static List<int[]> chicken, home, q;
	static Scanner sc = new Scanner(System.in);

	static void input() {
		n = sc.nextInt();
		m = sc.nextInt();
		chicken = new ArrayList<int[]>();
		home = new ArrayList<int[]>();
		q = new ArrayList<int[]>();
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				int a = sc.nextInt();
				if (a == 2) {
					chicken.add(new int[] { i, j });
				}
				if (a == 1) {
					home.add(new int[] { i, j });
				}
			}
		}
		selected = new boolean[chicken.size()];
	}

	static void go() {
		int sum = 0;
		for (int i = 0; i < home.size(); i++) {
			int dist = (int) 2e9;
			for (int j = 0; j < q.size(); j++) {
				int x = q.get(j)[0], y = q.get(j)[1];
				dist = Math.min(dist, Math.abs(x - home.get(i)[0]) + Math.abs(y - home.get(i)[1]));
			}
			sum += dist;
		}
		ans = Math.min(ans, sum);
	}

	static void combination(int idx, int cnt) {
		if (cnt == m) {
			q.clear();
			for (int i = 0; i < chicken.size(); i++) {
				if (selected[i]) {
					int x = chicken.get(i)[0], y = chicken.get(i)[1];
					q.add(new int[] { x, y });
				}
			}
			go();
			return;
		}
		for (int i = idx; i < chicken.size(); i++) {
			if (!selected[i]) {
				selected[i] = true;
				combination(i + 1, cnt + 1);
				selected[i] = false;
			}
		}
	}

	public static void main(String[] args){
		input();
		combination(0, 0);
		System.out.println(ans);
	}
}